wsgi.py

```diff
@@ -0,0 +1,72 @@
+import logging
+import os
+import sys
+import threading
+
+from flask import Flask, jsonify
+
+
+from gunicorn.arbiter import Arbiter
+
+
+gunicorn_logger = logging.getLogger('gunicorn.error')
+
+# all gunicorn processes in a given instance need to access a common
+# folder in /tmp where the metrics can be recorded
+PROMETHEUS_MULTIPROC_DIR = '/tmp/aiops_incoming_listener'
+
+try:
+    os.makedirs(PROMETHEUS_MULTIPROC_DIR, exist_ok=True)
+    os.environ['prometheus_multiproc_dir'] = PROMETHEUS_MULTIPROC_DIR
+    import kafka_app as APP
+except IOError as e:
+    # this is a non-starter for scraping metrics in the
+    # Multiprocess Mode (Gunicorn)
+    # terminate if there is an exception here
+    gunicorn_logger.error(
+        "Error while creating prometheus_multiproc_dir: %s", e
+    )
+    sys.exit(Arbiter.APP_LOAD_ERROR)
+
+
+# W0212 Access to a protected member _conns of a client class [pylint]
+# pylint: disable=W0212
+
+
+def create_application():
+    """Create App."""
+    app = Flask(__name__)
+    thread = threading.Thread(target=APP.main)
+    thread.start()
+
+    return app
+
+
+application = create_application()
+
+
+@application.route("/", methods=['GET'])
+def get_root():
+    """Root Endpoint."""
+    if APP.KAFKA_RESOURCES.get('consumer') and \
```

Comment:
So all my testing for `APP.KAFKA_RESOURCES.get('consumer')._client._conns` was done locally (since we can't control the strimzi kafka server on the dev cluster)

When I bring the server down, `APP.KAFKA_RESOURCES.get('consumer')._client._conns` is an empty dict; and when I bring the server up, it contains the connection object -- it works really great, it is a very dependable method to determine if the server is up/down, and I wish `aiokafka` had an API for this.

With `.subscription() ` however, the return value is always the list of currently subscribed topics despite the server being down -- this is with local testing ...although I do not know how it will work in the dev cluster, and hence cannot say with guarantee that this will/won't work.

Reply:
I see. Let's ditch the `.subscription` then.

I'm still not sure about the effect it can have here. What I was able to understand from aiokafka docs and sources, the `_conns` dict holds current connections to nodes. But if the connection breaks, it is [removed](https://github.com/aio-libs/aiokafka/blob/9756b332eb650163819675fd327b2f0992564b27/aiokafka/client.py#L391) and [reestablished](https://github.com/aio-libs/aiokafka/blob/9756b332eb650163819675fd327b2f0992564b27/aiokafka/client.py#L421) right after that.

The problem I have here is that aiokafka's internal mechanism can handle the reconnecting issue and I'm not sure if marking the listener service as not ready at this point is the appropriate answer. Because we would overreact to a simple reconnection. Also with multi-node Kafka deployment, these results are not reliable, because each node has a different role in the cluster. So our client can have `_conns` values for certain nodes but may be missing connection for example to a coordinator or to one of the partitions. And we wouldn't know that. 

So, what about:
```python
# instead of
KAFKA_RESOURCES = {}
...
KAFKA_RESOURCES['consumer'] = consumer
# let's have
KAFKA_CLIENT = consumer._client
```

An now (I would wrap this as a function in the kafka app):
```python
def is_online():
    ready = MAIN_LOOP.run_until_complete(asyncio.gather(
        KAFKA_CLIENT.ready(n.nodeId) for n in KAFKA_CLIENT.cluster.brokers()
    ))
    if all(ready):
        return True
    return False
```
^ was not tested, writing from top of my head. In words: see what nodes are available, and ask if they are all ready.

Reply:
> The problem I have here is that aiokafka's internal mechanism can handle the reconnecting issue and I'm not sure if marking the listener service as not ready at this point is the appropriate answer. Because we would overreact to a simple reconnection. 

I would not worry about that because both liveness and readiness probes are triggered continuously by OpenShift after a certain time interval. This time interval value is set with `periodSeconds` 
So assuming that listener fails the liveness/readiness check once because kafka is reconnecting, there is always a next time when we would check for `_conns`...and a next time...and a next time ...until it hits the `timeoutSeconds` value

Also, the `_conns` value seems pretty reliable to me..
Locally on my machine it is showing as -
```py
{(0, 0): <AIOKafkaConnection host=cluster-name-kafka-bootstrap port=9092>}
```

which is the very basic connection to kafka, and if that does not exist, nothing else matters.

Reply:
1. Fair point about the interval on the probes and multiple checks.
2. I don't think we can assume the connection count to be constant. The Kafka can loadbalance and spin new nodes/shut down unused ones.

Sigh, I don't like it but let's go with the simple presence of connections in the `._conns` then... Let's change the `KAFKA_RESOURCES['consumer']` to `KAFKA_CLIENT` for simplicity and merge it :slightly_smiling_face:

Reply:
thanks!

Will work on the Flask server tests now.

Reply:
`KAFKA_RESOURCES` is a global that needs to be some kind of a data structure to hold the value of `consumer` if/when it becomes available.
Simply using `KAFKA_CLIENT` would make it a global constant, and won't work, right?

Reply:
You're right. Simple usage of `KAFKA_CLIENT` wouldn't work, because when you'd assign the `KAFKA_CLIENT = consumer._client` this would create a local variable in the function. So the flask (or anywhere outside of the function) wouldn't know about this variable. It would be `None` for everyone else. When you are changing a global variable you need to use [the `global` keyword](https://docs.python.org/3/reference/simple_stmts.html#global). This way you can modify the global variable.

The reason it works with the dictionary is that you're not changing the reference the `KAFKA_RESOURCES` is pointing to. Since dictionaries are mutable, you're just changing the content. :wink: 

Also, Python is a pass-by-reference lang. So nothing needs to hold a value, it will hold the reference only. But since some objects are mutable, it can give a false impression that at some cases it can pass by value. It's not the case though.

Therefore you should use:
```python
KAFKA_CLIENT = None
...
async def consume_messages() -> None:
    ...
    global KAFKA_CLIENT
    KAFKA_CLIENT = consumer._client
    ...
```

Reply:
Thanks @tumido, that's good information about Python being pass-by-reference lang.

> Since dictionaries are mutable, you're just changing the content.

Yes, precisely! I needed a global-mutable combination as opposed to a global-constant, so Dictionary worked out perfectly! :)


Although my initial plan, in fact, *was* to use a global variable for the `consumer` object which would have required me to use the `global` keyword.  
`dict` gets the job done without having to use the `global` keyword, so I figured that would be a better implementation.

`dict` approach just seems better to me...
Can you re-eval both approaches one more time and if you still prefer `global KAFKA_CLIENT`, we can surely go with that! :)

Reply:
Despite both of these approaches work, I think having a global constant and changing it with the `global` is more understandable - [you're clearly stating the side effect](https://docs.python.org/3/faq/programming.html#what-are-the-rules-for-local-and-global-variables-in-python).

Using a global mutable object is not wrong. But... (and now it gets more philosophical :smile: ) It's a dictionary. Dictionary is intended as a map/hash table. What's the point of having a hash table when there's no intention for it to have more than one key? Normally you would use a variable for a single value. Not a dict. But now, you've chosen to use a dict just because of the mutable property it has. That's kinda a hack... We're talking about intentions here now. Like.. of course I can store every variable I have in a `dict` of a `list` but why would I do that. I have single value, single object, so I should use the mechanism designed for single values - simple assignment.

If there's a reason to use `dict`, other than abusing it's mutability, I would say let's go with a dict. It's a hack to use dict to modify global variable. Clever, but unnecessary. I don't see a reason for additional complexity.

Reply:
> Using a global mutable object is not wrong

Well, certainly not wrong -- it's mostly a personal preference I would say...
I do not like the explicit use of the word "global", just a personal preference you see..
(apparently pylama also does not like it :) )

Also I do not consider use of `dict` as a hack/abuse. For me, it is  "leveraging" the mutability! 😄  There are probably a few more interesting keys from `consumer` that theoretically we could have stored in the dict... we chose not to.

Anyway, let's go with `global KAFKA_CLIENT` for now, hopefully it will start growing on me 😉

Reply:
I don't think there's a reason not to use `global` statement. It is present in the lang syntax precisely for this occasions, when you want to modify global scope. Pylama doesn't like it and raises concern because modifying global scope is not encouraged practice, yet it's sometimes necessary (like in this case). Moreover it supports the argument of explicitness - it's clearly visible what we're doing on that line. Also it was a great exercise and journey into Python internals. At least i hope it was educational for you as it was for me. :wink:  

I think both approaches would be valid, and I would incline towards the dict usage, if we choose to store more complex structures, multiple properties etc. But we don't need that.

Anyhow, I think this issue is marginal and overall this is a good PR. Let's merge it. :+1: 
I've just moved the `pylint disable` to closer scope, so it's not disabled globally.

Reply:
@AparnaKarve I think this is a good read on this topic :wink: 

https://stackoverflow.com/a/14505529

Reply:
@tumido Yes indeed, that's a good read!